substrate/network-libp2p/src/network_state.rs

```diff
@@ -0,0 +1,796 @@
+						Ok(())
+					}
+					Err(err) => {
+						warn!(target: "sub-libp2p", "Failed to flush changes to JSON \
+													peer store: {}", err);
+						Err(err)
+					}
+				}
+			}
+		}
+	}
+}
+
+impl Drop for NetworkState {
+	fn drop(&mut self) {
+		let _ = self.flush_caches_to_disk();
+	}
+}
+
+// Assigns a `PeerId` to a node, or returns an existing ID if any exists.
+//
+// The function only accepts already-locked structs, so that we don't risk any deadlock.
+fn accept_connection(connections: &mut Connections, next_peer_id: &atomic::AtomicUsize,
+					 node_id: PeerstorePeerId, endpoint: Endpoint)
+	-> Result<PeerId, IoError>
+{
+	let peer_by_nodeid = &mut connections.peer_by_nodeid;
+	let info_by_peer = &mut connections.info_by_peer;
+
+	let peer_id = *peer_by_nodeid.entry(node_id.clone()).or_insert_with(|| {
+		let new_id = next_peer_id.fetch_add(1, atomic::Ordering::Relaxed);
+
+		let (tx, rx) = mpsc::unbounded();
+		let rx = rx
+			.into_future()
+			.map_err(|_| -> IoError { unreachable!() })
+			.and_then(|i| i.0.ok_or(IoError::new(IoErrorKind::ConnectionAborted, "kad aborted")));
+		let kad_connec = Box::new(rx) as Box<Future<Item = _, Error = _>>;
+
+		info_by_peer.insert(new_id, PeerConnectionInfo {
+			senders: Vec::new(),    // TODO: Vec::with_capacity(num_registered_protocols),
+			opened_kad: false,
+			incoming_kad_channel: tx,
+			kad_connec: kad_connec.shared(),
+			id: node_id.clone(),
+			originated: endpoint == Endpoint::Dialer,
+			ping: Mutex::new(None),
+			client_version: None,
+			local_address: None,
+			remote_address: None,
+		});
+		new_id
+	});
+
+	Ok(peer_id)
+}
+
+// Parses an address of the form `/ip4/x.x.x.x/tcp/x/p2p/xxxxxx`, and adds it to the given
+// peerstore. Returns the corresponding peer ID.
+fn parse_and_add_to_peerstore(addr_str: &str, peerstore: &PeersStorage)
+	-> Result<PeerstorePeerId, Error>
+{
+	let mut addr: Multiaddr = addr_str.parse().map_err(|_| ErrorKind::AddressParse)?;
+	let p2p_component = addr.pop().ok_or(ErrorKind::AddressParse)?;
+	let peer_id = match p2p_component {
+		AddrComponent::P2P(key) | AddrComponent::IPFS(key) => {
+			PeerstorePeerId::from_bytes(key).map_err(|_| ErrorKind::AddressParse)?
+		}
+		_ => return Err(ErrorKind::BadProtocol.into()),
+	};
+
+	// Registering the bootstrap node with a TTL of 100000 years   TODO: wrong
+	match peerstore {
+		PeersStorage::Memory(ref peerstore) => {
+			peerstore
+				.peer_or_create(&peer_id)
+				.add_addr(addr, Duration::from_secs(100000 * 365 * 24 * 3600));
+		}
+		PeersStorage::Json(ref peerstore) => {
+			peerstore
+				.peer_or_create(&peer_id)
+				.add_addr(addr, Duration::from_secs(100000 * 365 * 24 * 3600));
+		}
+	}
+	
+	Ok(peer_id)
+}
+
+// Obtains or generates the local private key using the configuration.
+fn obtain_private_key(config: &NetworkConfiguration) -> Result<secio::SecioKeyPair, IoError> {
+	Ok(if let Some(ref secret) = config.use_secret {
```

Comment:
I'd generally shy away from an `Ok(` wrapper of an `if` block once the block becomes more than 3 indent levels deep. a simple `let r = ...; Ok(r)` is more readable